### PR TITLE
Refactor scoring PoWs and deprecate old hash functions

### PR DESCRIPTION
**Pull request by grtlr**
# Description of change

This is the first step of bringing `bee-protocol` over to the `dev` branch. By storing the hash functions in `ProofOfWork`, we can call `ProofOfWork::hash` multiple times without performance hits. The next step would be to use `bee-pow` in the `chrysalis-pt2` branch.

A nice benefit would be that the hashing rate would be easier to benchmark, because currently we need to rebuild the complete node if we change small pieces in the `Hasher`.

## Links to any relevant issues

This PR is a first step towards #829.

## Type of change

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Ran all test cases.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Review by pvdrz — DISMISSED** [review]

**Comment by grtlr** [comment]
If we merge this, it might make sense to push that version to `crates.io` so that potential users see the deprecation of these methods.

File: bee-pow/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,8 @@ homepage = "https://www.iota.org"
 bee-crypto = { version = "0.2.1-alpha", path = "../bee-crypto", default-features = false }
 bee-ternary = { version = "0.5.0", path = "../bee-ternary", default-features = false }
 
-iota-crypto = { version = "0.7.0", default-features = false, features = [ "blake2b", "digest" ] }
+# Wait for new release of `iota-crypto` that uses `bee-node:0.5.0`.
```
Comment (pvdrz) [inline bee-pow/Cargo.toml]: Does this block the PR?

**Review by pvdrz — DISMISSED** [review]

File: bee-pow/Cargo.toml (hunk 1 above)
Comment (grtlr) [reply bee-pow/Cargo.toml]: Now that @thibault-martinez has bumped the version of `iota-crypto` 💪 everything is fine.

File: bee-crypto/Cargo.toml [hunk 2]
```diff
@@ -20,6 +20,7 @@ tiny-keccak = { version = "2.0.2", default-features = false, features = [ "kecca
 
 [dev-dependencies]
 criterion = { version = "0.3.5", default-features = false }
+iota-crypto = { version = "0.8.0", default-features = false, features = [ "blake2b", "digest", "curl-p"] }
```
Comment (thibault-martinez) [inline bee-crypto/Cargo.toml]:
```suggestion
iota-crypto = { version = "0.8.0", default-features = false, features = [ "blake2b", "digest", "curl-p" ] }
```

Comment (grtlr) [reply bee-crypto/Cargo.toml]: Oops :)

**Comment by grtlr** [comment]
We should prepare the corresponding PR on `chrysalis-pt2` before we merge this, otherwise our CI might start failing there.

File: bee-signing/tests/wots_sponge.rs [hunk 3]
```diff
@@ -1,7 +1,10 @@
 // Copyright 2020-2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use bee_crypto::ternary::sponge::{CurlP27, CurlP81, Kerl, Sponge};
+#[allow(deprecated)]
+use bee_crypto::ternary::sponge::{CurlP27, CurlP81};
+#[allow(deprecated)]
+use bee_crypto::ternary::sponge::{Kerl, Sponge};
```
Comment (thibault-martinez) [inline bee-signing/tests/wots_sponge.rs]: Merge both imports

File: bee-signing/tests/ternary_seed.rs [hunk 4]
```diff
@@ -1,7 +1,9 @@
 // Copyright 2020-2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use bee_crypto::ternary::sponge::{Kerl, Sponge};
+#[allow(deprecated)]
+use bee_crypto::ternary::sponge::Kerl;
+use bee_crypto::ternary::sponge::Sponge;
```
Comment (thibault-martinez) [inline bee-signing/tests/ternary_seed.rs]: I think I would even deprecate Sponge tbh... The Curl version in iota-crypto doesn't implement this trait.

File: bee-pow/Cargo.toml [hunk 5]
```diff
@@ -14,7 +14,7 @@ homepage = "https://www.iota.org"
 bee-crypto = { version = "0.2.1-alpha", path = "../bee-crypto", default-features = false }
 bee-ternary = { version = "0.5.0", path = "../bee-ternary", default-features = false }
 
-iota-crypto = { version = "0.7.0", default-features = false, features = [ "blake2b", "digest" ] }
+iota-crypto = { version = "0.8.0", default-features = false, features = [ "blake2b", "digest", "curl-p"] }
```
Comment (thibault-martinez) [inline bee-pow/Cargo.toml]:
```suggestion
iota-crypto = { version = "0.8.0", default-features = false, features = [ "blake2b", "digest", "curl-p" ] }
```

File: bee-crypto/tests/sponge.rs [hunk 6]
```diff
@@ -1,7 +1,9 @@
 // Copyright 2020-2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use bee_crypto::ternary::sponge::{CurlP27, CurlP81, Kerl, Sponge, SpongeKind, UnrolledCurlP81};
+#[allow(deprecated)]
+use bee_crypto::ternary::sponge::{CurlP27, CurlP81, Kerl};
+use bee_crypto::ternary::sponge::{Sponge, SpongeKind, UnrolledCurlP81};
```
Comment (thibault-martinez) [inline bee-crypto/tests/sponge.rs]: If we deprecate all current sponge kinds, there is not reason not to deprecate `SpongeKind` ?

File: bee-crypto/CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Deprecated
 
+- Deprecated `CurlP` from `bee-crypto` (now uses `iota-crypto` instead)
+
```
Comment (thibault-martinez) [inline bee-crypto/CHANGELOG.md]: Can you actually prepare a release ? Maybe `0.3.0`, it's probably time to drop the alpha tag anyway.

File: bee-pow/CHANGELOG.md [hunk 8]
```diff
@@ -11,6 +11,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Scoring of Proof of Work can now reuse hash functions
```
Comment (thibault-martinez) [inline bee-pow/CHANGELOG.md]: Same, prepare a release, 0.2.0

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Comment by grtlr** [comment]
> We should prepare the corresponding PR on `chrysalis-pt2` before we merge this, otherwise our CI might start failing there.

I've checked, this PR _shouldn't_ cause any problems with `chrysalis-pt2`.

File: Cargo.toml [hunk 9]
```diff
@@ -27,3 +27,7 @@ panic = "abort"
 
 [profile.release]
 panic = "abort"
+
+# `iota-crypto` uses bee-ternary. We patch it to use our local version to avoid importing it twice.
+[patch.crates-io]
+bee-ternary = { path = "bee-ternary" }
```
Comment (grtlr) [inline Cargo.toml:33]: Maybe this can be changed again back when we start handling `Cargo.lock` files differently.

File: bee-message/Cargo.toml [hunk 10]
```diff
@@ -12,14 +12,14 @@ homepage = "https://www.iota.org"
...
-bee-pow = { version = "0.1.0", path = "../bee-pow", default-features = false }
+bee-pow = { version = "0.2.0", path = "../bee-pow", default-features = false }
 bee-ternary = { version = "0.5.2", path = "../bee-ternary", default-features = false, features = [ "serde1" ] }
 
 bech32 = { version = "0.8.1", default-features = false }
 bytemuck = { version = "1.7.2", default-features = false }
 digest = { version = "0.9.0", default-features = false }
 hex = { version = "0.4.3", default-features = false }
-iota-crypto = { version = "0.7.0", default-features = false, features = [ "ed25519", "blake2b" ] }
+iota-crypto = { version = "0.9.0", default-features = false, features = [ "ed25519", "blake2b", "std" ] }
```
Comment (grtlr) [inline bee-message/Cargo.toml]: Wait for #844 and https://github.com/iotaledger/crypto.rs/pull/118

```suggestion
iota-crypto = { version = "0.9.1", default-features = false, features = [ "ed25519", "blake2b" ] }
```

File: bee-signing/CHANGELOG.md [hunk 11]
```diff
@@ -15,12 +15,20 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Removed
 
-PrivateKey: Drop bound
-
 ### Fixed
 
 ### Security -->
 
+## 0.2.0 - 2021-11-17
```
Comment (thibault-martinez) [inline bee-signing/CHANGELOG.md]: Change date (hopefully today, should be fine)

File: bee-pow/CHANGELOG.md [hunk 12]
```diff
@@ -19,6 +19,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 0.2.0 - 2021-11-17
```
Comment (thibault-martinez) [inline bee-pow/CHANGELOG.md]: Same

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: Cargo.lock [hunk 13]
```diff
@@ -1479,13 +1479,15 @@ dependencies = [
 
 [[package]]
 name = "iota-crypto"
-version = "0.7.0"
+version = "0.9.0"
```
Comment (thibault-martinez) [inline Cargo.lock]: Any idea why this is not 0.9.1 ?

File: bee-crypto/Cargo.toml [hunk 14]
```diff
@@ -13,6 +13,7 @@ homepage = "https://www.iota.org"
 [dependencies]
 bee-ternary = { version = "0.5.2", path = "../bee-ternary", default-features = false }
 
+iota-crypto = { version = "0.9.1", default-features = false, features = [ "curl-p" ] }
```
Comment (thibault-martinez) [inline bee-crypto/Cargo.toml]: Alphanumeric order please :)

File: bee-crypto/CHANGELOG.md [hunk 15]
```diff
@@ -19,6 +19,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 0.3.0 - 2021-11-17
```
Comment (thibault-martinez) [inline bee-crypto/CHANGELOG.md]: Same

**Review by thibault-martinez — DISMISSED** [review]

**Review by pvdrz — DISMISSED** [review]

File: bee-pow/src/score.rs [hunk 16]
```diff
@@ -3,43 +3,81 @@
...
+            .iter()
+            .for_each(|t| self.pow_input.push(t));
+        b1t6::encode::<T1B1Buf>(tail)
+            .iter()
+            .for_each(|t| self.pow_input.push(t));
+
+        // Pad to 243 trits
+        self.pow_input.push(Btrit::Zero);
+        self.pow_input.push(Btrit::Zero);
+        self.pow_input.push(Btrit::Zero);
```
Comment (Alex6323) [inline bee-pow/src/score.rs]: I think you could also use the `append` API here on `TritBuf`:

```
self.pow_input.append(b1t6::encode::<T1B1buf>(&pow_digest).as_slice());
self.pow_input.append(b1t6::encode::<T1B1buf>(tail).as_slice());
```
Maybe to reduce repetition also create a slice of the 3 `Btrit::Zero` and append it as well. :slightly_smiling_face:

**Review by Alex6323 — DISMISSED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**